### PR TITLE
Fixing Windows Support

### PR DESCRIPTION
**Pull request by parthsane**
Made changes which solves some dependency issues in sgx_detect,
  but it still needs more work

  Currently sgx_detect and rs-libc don't compile on windows

### CODE REVIEW

File: rs-libc/build.rs [hunk 1]
```diff
@@ -42,6 +42,7 @@ fn main() {
         "librsc.a"
     };
 
+
     build
         .define(
             "weak_alias(old,new)",
```
Comment (parthsane) [inline rs-libc/build.rs]: -ffreestanding is not supported by MSVC, I am unable to find the correct substitute.

File: sgxs-tools/src/sgx_detect/main.rs [hunk 2]
```diff
@@ -232,9 +235,13 @@ impl SgxSupport {
             Ok(TimeoutHardError::new(client))
```
Comment (parthsane) [inline sgxs-tools/src/sgx_detect/main.rs]: On line 219 to 224, there is use of certain functions which depend on Linux specific functions.

File: rs-libc/build.rs [hunk 3]
```diff
@@ -42,6 +42,7 @@ fn main() {
         "librsc.a"
     };
 
+
```
Comment (jethrogb) [inline rs-libc/build.rs]: stray newline

File: rs-libc/build.rs (hunk 1 above)
Comment (jethrogb) [reply rs-libc/build.rs]: Take a look at https://wiki.osdev.org/Visual_Studio#Compiler_Options and see what would be good. At least `/NODEFAULTLIB` I'd think

File: rs-libc/src/c/floatscan.c [hunk 4]
```diff
@@ -418,7 +417,9 @@ static long double hexfloat(FILE *f, int bits, int emin, int sign, int pok)
 	return scalbnl(y, e2);
 }
 
+#ifndef _MSC_VER
 __attribute__ ((visibility ("hidden")))
```
Comment (jethrogb) [inline rs-libc/src/c/floatscan.c]: Yeah I think unfortunately there is just no equivalent of this in MSVC. So this ifdef is fine.

File: rs-libc/src/c/strcpy.c [hunk 5]
```diff
@@ -1,8 +1,14 @@
 #include <string.h>
 
+#ifdef _MSC_VER
+#define RESTRICT __restrict
+#else
+#define RESTRICT restrict
+#endif
```
Comment (jethrogb) [inline rs-libc/src/c/strcpy.c]: Instead of adding this block in every file, can you just do the equivalent of `-Drestrict=__restrict` on the command line?

File: sgxs-tools/Cargo.toml [hunk 6]
```diff
@@ -33,7 +33,7 @@ path = "src/sgx_detect/main.rs"
 "sgxs-loaders" = { version = "0.2.0", path = "../sgxs-loaders" }
 "aesm-client" = { version = "0.2.0", path = "../aesm-client", features = ["sgxs"] }
 "sgx-isa" = { version = "0.3.0", path = "../sgx-isa" }
-"dcap-ql" = { version = "0.2.0", path = "../dcap-ql" }
+# "dcap-ql" = { version = "0.2.0", path = "../dcap-ql" }
```
Comment (jethrogb) [inline sgxs-tools/Cargo.toml]: Please make this a [Linux-specific dependency](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#platform-specific-dependencies)

File: sgxs-tools/src/sgx_detect/main.rs (hunk 2 above)
Comment (jethrogb) [reply sgxs-tools/src/sgx_detect/main.rs]: Make a copy of `linux.rs` for windows, and implement Windows functionality there. I think for `rdmsr`, `aesm_status`, `kmod_status` you can always return an error for now (although something should be possible for aesm_status & kmod_status). For read_efi_var you can do [GetFirmwareEnvironmentVariable](https://docs.rs/winapi/0.3.8/x86_64-pc-windows-msvc/winapi/um/winbase/fn.GetFirmwareEnvironmentVariableA.html).

File: rs-libc/build.rs (hunk 1 above)
Comment (parthsane) [reply rs-libc/build.rs]: `cc` doesn't pass the flag `/NODEFAULTLIB` to the `link.exe`

**Comment by parthsane** [comment]
@jethrogb I think this is ready to be reviewed. I changed the name of the PR just to be clear

**Comment by jethrogb** [comment]
Can you please rebase on master?

File: sgxs-tools/src/sgx_detect/imp/windows.rs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+use failure::{Error, Fail, ResultExt};
+
+use crate::DetectError;
+use crate::interpret::{AesmStatus, KmodStatus};
+use winapi::um::winbase::GetFirmwareEnvironmentVariableA;
+use winapi::shared::minwindef::DWORD;
+use winapi::um::winnt::{LPCSTR, PVOID};
+
+pub fn rdmsr(address: u64) -> Result<u64, Error> {
+    bail!("Currently not implemented")
```
Comment (jethrogb) [inline sgxs-tools/src/sgx_detect/imp/windows.rs]: Be more specific, such as:

> RDMSR not implemented on Windows

Same thing below

File: sgxs-tools/src/sgx_detect/imp/windows.rs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        GetFirmwareEnvironmentVariableA(
+            name.as_ptr() as _,
+            guid.as_ptr() as _,
+            env.as_mut_ptr() as _,
+            env.len() as _
+        )
+    };
+    if ret == 0 {
+        return Err(DetectError::EfiVariableError(std::io::Error::last_os_error()).into())
+    }
```
Comment (jethrogb) [inline sgxs-tools/src/sgx_detect/imp/windows.rs]: You should truncate `env` if `ret` is not 0

File: sgxs-tools/src/sgx_detect/imp/windows.rs [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+pub fn rdmsr(address: u64) -> Result<u64, Error> {
+    bail!("Currently not implemented")
+}
+
+pub fn read_efi_var(name: &str, guid: &str) -> Result<Vec<u8>, Error> {
+    let mut env: Vec<u8> = vec![0; 1024];
+    let ret: DWORD = unsafe {
+        GetFirmwareEnvironmentVariableA(
+            name.as_ptr() as _,
+            guid.as_ptr() as _,
```
Comment (jethrogb) [inline sgxs-tools/src/sgx_detect/imp/windows.rs]: The types of these parameters are `LPCSTR`. So you need to convert the Rust strings to C strings before passing them. Use `std::ffi::CString`.

File: aesm-client/Cargo.toml [hunk 10]
```diff
@@ -26,8 +26,7 @@ test-sgx = []
 
 [dependencies]
 # Project dependencies
-sgxs = { version = "0.7.0", path = "../sgxs", optional = true }
-
+sgxs = { version = "0.7.0", path = "../sgxs"}
```
[the remark this answers is not in the document]
Comment (parthsane) [reply aesm-client/Cargo.toml]: Hmm. This is clearly wrong. I should definitely remove this for unix at least.

The biggest issue is that the sgx_uae_service api on Windows needs `sgxs::sigstruct::{Attributes, Sigstruct}` (actually it wants the pointer to a C style struct representing them). I don't quiet have an idea as to how to make this crate optional for windows while still having access to the definition of `sgxs::sigstruct::{Attributes, Sigstruct}`

[the remark this answers is not in the document]
Comment (jethrogb) [reply aesm-client/Cargo.toml]: Those types are defined in sgx-isa

File: Cargo.toml [hunk 11]
```diff
@@ -14,4 +14,16 @@ members = [
     "sgxs-loaders",
     "sgxs-tools"
 ]
+default-members = [
```
Comment (jethrogb) [inline Cargo.toml]: Now that not everything is built by default, CI should be modified so the excluded packages are still tested. Or, you can just not set this.

File: aesm-client/src/imp/windows.rs [hunk 12]
```diff
@@ -1,314 +1,119 @@
...
+type SgxInitQuoteFn = unsafe extern "system" fn(
+    p_target_info: *mut Targetinfo,
+    p_gid: *mut u8,
+) -> u32;
+
+struct Library {
+    library: dl::Library,
+    pub init_quote: SgxInitQuoteFn,
+    pub get_quote: SgxGetQuoteFn,
+    pub get_launch_token: GetLaunchTokenFn,
```
Comment (jethrogb) [inline aesm-client/src/imp/windows.rs]: I don't think these fields need to be `pub`.

File: aesm-client/src/imp/windows.rs [hunk 13]
```diff
@@ -1,314 +1,119 @@
...
             }
-            CoUninitialize();
         }
     }
 }
 
 impl AesmClient {
     pub fn new() -> Self {
-        Default::default()
+        AesmClient::default()
```
Comment (jethrogb) [inline aesm-client/src/imp/windows.rs]: nit: unnecessary change

File: aesm-client/src/imp/windows.rs [hunk 14]
```diff
@@ -1,314 +1,119 @@
...
-                }
-                CoUninitialize();
-                IoError::new(
-                    ErrorKind::Other,
-                    format!("Fail to create Aesm Interface {}", e),
-                )
-            })?;
-        }
-        Ok(AesmInterface { inner: interface })
+    fn get_library(&self) -> Arc<Library> {
```
Comment (jethrogb) [inline aesm-client/src/imp/windows.rs]: I don't think this function is necessary. You can just write `&self.library` where-ever you use this.

File: rs-libc/src/c/__ctype_b_loc.c [hunk 15]
```diff
@@ -1,9 +1,14 @@
...
+#ifdef _MSC_VER
+    #define __BYTE_ORDER __BYTE_ORDER__
+    #define __BIG_ENDIAN __ORDER_BIG_ENDIAN__
+#else
+    #include <endian.h>
+#endif
 
-#if __BYTE_ORDER == __BIG_ENDIAN
-#define X(x) x
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
```
Comment (jethrogb) [inline rs-libc/src/c/__ctype_b_loc.c]: This line shouldn't be changed. It now no longer works on Linux

File: sgxs-tools/src/sgx_detect/imp/windows.rs [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+use winapi::shared::minwindef::DWORD;
+use std::ffi::CString;
+
+pub fn rdmsr(_address: u64) -> Result<u64, Error> {
+    bail!("RDMSR not implemented on Windows")
+}
+
+pub fn read_efi_var(name: &str, guid: &str) -> Result<Vec<u8>, Error> {
+    let mut env: Vec<u8> = vec![0; 1024];
+    let name = CString::new(name)?.as_ptr();
```
Comment (jethrogb) [inline sgxs-tools/src/sgx_detect/imp/windows.rs]: You should store the `CString` in the variable and call `as_ptr` below when calling the C function. The way you've written it now, the `CString` is deallocated at the end of the statement.

File: aesm-client/src/imp/unix.rs [hunk 17]
```diff
@@ -149,10 +150,13 @@ impl AesmClient {
...
         &self,
-        mr_enclave: Vec<u8>,
-        signer_modulus: Vec<u8>,
-        attributes: Vec<u8>,
+        sigstruct: &Sigstruct,
+        attributes: Attributes,
     ) -> Result<Vec<u8>> {
+        let mr_enclave: Vec<u8> = sigstruct.enclavehash.to_vec();
+        let signer_modulus: Vec<u8> = sigstruct.modulus.to_vec();
+        let attributes: Vec<u8> = attributes.as_ref().to_vec();
```
Comment (jethrogb) [inline aesm-client/src/imp/unix.rs]: Type annotations are unnecessary, but also you can just write `sigstruct.enclavehash.to_vec()` etc. directly where it's being used below.

File: aesm-client/src/imp/windows.rs [hunk 18]
```diff
@@ -1,248 +1,89 @@
...
+            .field("library", &self.library)
+            .field("init_quote", &(self.init_quote as *const c_void))
+            .field(
+                "get_quote",
+                &(self.get_quote as *const c_void),
+            )
+            .field(
+                "get_launch_token",
+                &(self.get_launch_token as *const c_void),
+            )
```
Comment (jethrogb) [inline aesm-client/src/imp/windows.rs]: I think you should be able to put the whole `.field(...)` call on one line

File: aesm-client/src/imp/windows.rs [hunk 19]
```diff
@@ -253,62 +94,22 @@ impl AesmClient {
...
-                    target_info.len() as _,
-                    gid.as_mut_ptr(),
-                    gid.len() as _,
-                    &mut error as _,
-                )
-                .into_io_error()?;
-                if error != 0 {
-                    return Err(Error::AesmCode(error.into()));
-                }
+            let lib = &self.library;
```
Comment (jethrogb) [inline aesm-client/src/imp/windows.rs]: Don't need this assignment, can just use directly

File: aesm-client/src/imp/windows.rs [hunk 20]
```diff
@@ -323,69 +124,54 @@ impl AesmClient {
...
-        let mut error: AesmError = 0;
 
         unsafe {
-            let interface = self.create_instance()?;
-            if let Some(get_quote) = (*(*interface.inner).vtbl).get_quote {
-                get_quote(
-                    interface.inner,
-                    report.as_ptr(),
-                    report.len() as _,
+            let lib = &self.library;
```
Comment (jethrogb) [inline aesm-client/src/imp/windows.rs]: Don't need this assignment, can just use directly

File: aesm-client/src/imp/windows.rs [hunk 21]
```diff
@@ -323,69 +124,54 @@ impl AesmClient {
...
+
+            let (sig_rl_in, sig_rl_size_in )= match sig_rl.len() {
```
Comment (jethrogb) [inline aesm-client/src/imp/windows.rs]: ` )=` should be `) =`

File: aesm-client/src/imp/windows.rs [hunk 22]
```diff
@@ -323,69 +124,54 @@ impl AesmClient {
...
+                0 => (std::ptr::null(), 0),
+                s => (sig_rl.as_ptr(), s)
+            };
+
+            let error = (lib.get_quote)(
```
Comment (jethrogb) [inline aesm-client/src/imp/windows.rs]: You should assert that `report` and `spid` are of the appropriate size before passing them to C

File: aesm-client/src/imp/windows.rs [hunk 23]
```diff
@@ -323,69 +124,54 @@ impl AesmClient {
...
+                    sig_rl_in,
+                    sig_rl_size_in as _,
+                    qe_report.as_mut_ptr() as _,
+                    quote.as_mut_ptr() as _,
                     quote_buffer_size,
-                    &mut error as _,
-                )
-                .into_io_error()?;
+                );
                 if error != 0 {
```
Comment (jethrogb) [inline aesm-client/src/imp/windows.rs]: This statement should be indented less

File: aesm-client/src/imp/windows.rs [hunk 24]
```diff
@@ -323,69 +124,54 @@ impl AesmClient {
...
-                    licence_token.as_mut_ptr(),
-                    licence_token.len() as _,
-                    &mut error as _,
-                )
-                .into_io_error()?;
-                if error != 0 {
-                    return Err(Error::AesmCode(error.into()));
-                }
+
+            let lib = &self.library;
```
Comment (jethrogb) [inline aesm-client/src/imp/windows.rs]: Don't need this assignment, can just use directly

File: rs-libc/build.rs [hunk 25]
```diff
@@ -13,7 +13,10 @@ use std::path::PathBuf;
 fn main() {
     let arch = env::var("CARGO_CFG_TARGET_ARCH").unwrap();
     let p_c = ["src", "c"].iter().collect::<PathBuf>();
+    #[cfg(unix)]
     let p_s = ["src", "asm", &arch].iter().collect::<PathBuf>();
+    #[cfg(windows)]
+    let p_s = ["src", "obj", &arch].iter().collect::<PathBuf>();
```
Comment (jethrogb) [inline rs-libc/build.rs]: I think the directory should be `obj/windows/$ARCH`

File: sgxs-tools/src/sgx_detect/main.rs [hunk 26]
```diff
@@ -312,6 +327,7 @@ fn main() {
     if args.is_present("EXPORT") {
         serde_yaml::to_writer(io::stdout(), &support).unwrap();
         println!("");
+
```
Comment (jethrogb) [inline sgxs-tools/src/sgx_detect/main.rs]: Stray newline
